lib/guest_installation_and_configuration_base.pm

```diff
@@ -0,0 +1,1565 @@
+        $self->modify_guest_params($self->{guest_name}, 'guest_metadata_options') if (($self->{guest_installation_result} eq 'PASSED') and ($_current_metadata_options ne $self->{guest_metadata_options}));
+    }
+    return $self;
+}
+
+#Configure [guest_vcpus_options].User can still change [guest_vcpus] by passing non-empty arguments using hash.
+#If installations already passes,modify_guest_params will be called to modify [guest_vcpus] using already modified [guest_vcpus_options].
+sub config_guest_vcpus {
+    my $self = shift;
+
+    $self->reveal_myself;
+    my $_current_vcpus_options    = $self->{guest_vcpus_options};
+    my $_current_cpumodel_options = $self->{guest_cpumodel_options};
+    $self->config_guest_params(@_) if (scalar(@_) gt 0);
+    $self->{guest_vcpus}            = 2 if ($self->{guest_vcpus} eq '');
+    $self->{guest_vcpus_options}    = "--vcpus $self->{guest_vcpus}";
+    $self->{guest_cpumodel_options} = "--cpu $self->{guest_cpumodel}" if ($self->{guest_cpumodel} ne '');
+    $self->modify_guest_params($self->{guest_name}, 'guest_vcpus_options') if (($self->{guest_installation_result} eq 'PASSED') and ($_current_vcpus_options ne $self->{guest_vcpus_options}));
+    $self->modify_guest_params($self->{guest_name}, 'guest_cpumodel_options') if (($self->{guest_installation_result} eq 'PASSED') and ($_current_cpumodel_options ne $self->{guest_cpumodel_options}));
+    return $self;
+}
+
+#Configure [guest_memory_options].User can still change [guest_memory] by passing non-empty arguments using hash.
+#If installations already passes,modify_guest_params will be called to modify [guest_memory] using already modified [guest_memory_options].
+sub config_guest_memory {
+    my $self = shift;
+
+    $self->reveal_myself;
+    my $_current_memory_options = $self->{guest_memory_options};
+    $self->config_guest_params(@_) if (scalar(@_) gt 0);
+    $self->{guest_memory}         = 2048 if ($self->{guest_memory} eq '');
+    $self->{guest_memory_options} = "--memory $self->{guest_memory}";
+    $self->modify_guest_params($self->{guest_name}, 'guest_memory_options') if (($self->{guest_installation_result} eq 'PASSED') and ($_current_memory_options ne $self->{guest_memory_options}));
+    return $self;
+}
+
+#Configure [guest_virt_options].User can still change [host_hypervisor_uri],[host_virt_type] and [guest_virt_options] by passing non-empty arguments using hash.
+sub config_guest_virtualization {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->config_guest_params(@_) if (scalar(@_) gt 0);
+    if ($self->{guest_installation_result} eq '') {
+        $self->{guest_virt_options} = "--connect $self->{host_hypervisor_uri} " if ($self->{host_hypervisor_uri} ne '');
+        if ($self->{host_virt_type} eq '') {
+            $self->{host_virt_type} = 'kvm';
+            $self->{host_virt_type} = 'xen' if (script_output("zypper info xen_server | grep -iE \"installed.*yes\"", proceed_on_failure => 1) ne '');
```

Comment:
to determine if it is a xen host by checking xen_server installation is not accurate, we are used to install both xen_server and kvm_server in our test machines while boot up with one of them. So considering our testing purpose, suggest to check the kernel module loaded, 'lsmod | grep xen'.

Reply:
I think the most reliable way is:
journalctl -b | grep -i "Hypervisor detected: Xen"